ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/3_Manage_Complexity/3_2_Boolean_Calculator/src/Operator.ts

```diff
@@ -0,0 +1,97 @@
+import { Evaluateable } from "./Evaluateable";
+import { PartiallyParsedExpression } from "./ExpressionParser";
+
+export abstract class Operator extends Evaluateable {
+    static Parse(expression: PartiallyParsedExpression, OperatorClass: new (...params:any) => Operator): PartiallyParsedExpression{
+        return OperatorClass.prototype.parse(expression);
+    }
+
+    protected get name(): string {
+        return this.constructor.name.toUpperCase();
+    }
+
+    abstract evaluate(): boolean;
+    protected abstract parse(expression: PartiallyParsedExpression): PartiallyParsedExpression;
```

Comment:
I'm curious what do you think about this solution.
Using polymorphism and defining how each type of Operator is parsed in the class itself seemed like a good approach, so I could avoid ugly checks like [isUnary](https://github.com/enekesabel/the-software-essentialist/pull/6/commits/a5a6be624ae2b3b35154f9250ba11132bc063354#diff-e42bee31eb40784d4352a0628f93649920936f1c30cd38514c7dac790ee672c5R29).
However, this mixes the responsibility of parsing into the domain object. The other half of my brain says that the domain objects should know nothing about parsing, ExpressionParses should take on that responsibility.

I choose to stick with this solution, to me it looks like this makes the program more extensible. If I kept this part of the parsing logic in ExpressionParser, I'd possibly have to make some adjustments there, in case a new type of Operator (other than Unary or Binary) is introduced.

What do you think, what's the best way to solve problems like this?